deb and rpm packaging for all binaries

julienduchesne opened the pull request:
There is currently deb and rpm packages being published for logcli amd64 but they aren't signed

In this PR:
- Add RPM and deb packages for `logcli`, `loki-canary` `loki` and `promtail` for arm, arm64 and amd64
- Sign them with the Grafana GPG key (packages.grafana.com/gpg.key). This will be the same key for all Grafana products
- Add CI steps to test the RPM and deb packages. This launches a docker image with systemd installed, then it installs both loki and promtail and does a `logcli` query on the Loki instance

chaudum commented on tools/package-nfpm.jsonnet:
```diff
@@ -0,0 +1,41 @@
+  homepage: 'https://grafana.com/loki',
+  changelog: 'https://github.com/grafana/loki/blob/main/CHANGELOG.md',
+  license: 'AGPL-3.0',
+  contents: [{
+    src: './dist/tmp/packages/%s-linux-%s' % [name, arch],
+    dst: '/usr/local/bin/%s' % name,
+  }],
```
Did I understand it correctly that the package only contain the binary that is then installed to `/usr/local/bin`? This may be useful for LogCLI, but for Promtail and Loki, which need a configuration file, I see more harm than gain. Users would expect that the package would also contain a systemd unit which they can use to start/stop Loki and a config that works out of the box (probably using filesystem storage). Unless the packages do not contain that, I am worried to publish them.

jeschkies replied:
I agree. Configuration and systemd unit files are a big plus of native packages.

julienduchesne replied:
👍 Good point, I'll add those. Adding daemon configs really is the nicest part of OS packaging

chaudum reviewed (changes requested):
Thanks for your efforts to make OS packages happen also for Loki. However I have a big concern which I raised in a comment.

jeschkies commented on tools/package-nfpm.jsonnet:
```diff
@@ -0,0 +1,41 @@
+  vendor: 'Grafana Labs Inc',
+  homepage: 'https://grafana.com/loki',
+  changelog: 'https://github.com/grafana/loki/blob/main/CHANGELOG.md',
+  license: 'AGPL-3.0',
```
Promtail is Apache 2 if I'm not wrong.

julienduchesne replied:
I changed this

jeschkies reviewed (commented):
Thanks for getting this started :slightly_smiling_face:

KMiller-Grafana reviewed (approved):
File contents for tools/packaging/nfpm.jsonnet look good to me.

julienduchesne commented:
@jeschkies @chaudum. Re-requesting a review. I added systemd configs for both deb and RPM. I also added a CI test to check that everything keeps working. The CI check uses CentOS and Debian. It installs loki and promtail and then queries with logcli to see that everything works well

jeschkies commented on .drone/drone.jsonnet:
```diff
@@ -537,18 +539,93 @@ local manifest_ecr(apps, archs) = pipeline('manifest-ecr') {
+          }),
+      run('test packaging',
+          commands=[
+            'go install github.com/google/go-jsonnet/cmd/jsonnet@latest',  // Test, install in build image instead
```
I'm not sure I understand this step. Why do you install a Go dependency here?

julienduchesne replied:
Installed in the build image here: https://github.com/grafana/loki/pull/6526

jeschkies commented on .drone/drone.jsonnet:
```diff
@@ -537,18 +539,93 @@ local manifest_ecr(apps, archs) = pipeline('manifest-ecr') {
+      {
+        name: 'test deb package',
+        image: 'docker',
+        commands: ['./tools/packaging/verify-deb-install.sh'],
```
This is awesome. I'd use some test library but am totally fine with a shell script for now.

jeschkies commented on tools/packaging/nfpm.sh:
```diff
@@ -0,0 +1,24 @@
+for name in loki loki-canary logcli promtail; do
+    for arch in amd64 arm64 arm; do
+        config_path="dist/tmp/config-${name}-${arch}.json"
+        jsonnet -V "name=${name}" -V "arch=${arch}" "tools/packaging/nfpm.jsonnet" > "${config_path}"
```
...ah this is why you need to install Jsonnet :slightly_smiling_face:

jeschkies commented on tools/packaging/verify-deb-install.sh:
```diff
@@ -0,0 +1,34 @@
+
+cat <<EOF | docker exec --interactive "${image}" sh
+    # Install loki and check it's running
+    dpkg -i ${dir}/dist/loki_0.0.0~rc0_amd64.deb
```
Shouldn't the binary have a version?

julienduchesne replied:
oh dang, that'll fail on actual releases. Will fix

jeschkies commented on tools/packaging/verify-deb-install.sh:
```diff
@@ -0,0 +1,34 @@
+    dpkg -i ${dir}/dist/logcli_0.0.0~rc0_amd64.deb
+
+    # Check that there are labels
+    sleep 5
+    labels_found=\$(logcli labels)
+    echo "Found labels: \$labels_found"
+    [ "\$labels_found" != "" ] || (echo "no logs found with logcli" && exit 1)
```
I'd use a loop but this is a great start. We can always iterate.

jeschkies reviewed (approved):
This is really awesome! Just so I understand. The packages are not published yet, right?

julienduchesne commented:
> This is really awesome! Just so I understand. The packages are not published yet, right?

No, we're working on a global repository (apt.grafana.com and rpm.grafana.com) where we will publish all packages. It'll just be an additional Drone steps. For now, the deb and rpm files will be in the releases. I'm hoping to have the repositories ready before end of July

rightaway commented:
@julienduchesne The debian package will install directly on the system not through docker?

Where can I see what the `.service` files will look like on debian for loki and promtail packages?

chaudum commented on tools/packaging/nfpm.jsonnet:
```diff
@@ -0,0 +1,90 @@
+        dst: '/etc/systemd/system/loki.service',
+      },
+      {
+        src: './cmd/loki/loki-local-config.yaml',
```
This config file configures Loki to write to `/tmp/loki`. Not sure if this is the ideal default location, but I think it's good enough for a first iteration.

MAFLO321 commented on tools/packaging/promtail.service:
```diff
@@ -0,0 +1,11 @@
+Type=simple
+User=promtail
+ExecStart=/usr/bin/promtail -config.file /etc/promtail/config.yml
+
```
Wouldn't it be a neat idea if promtail would also be `Restart = on-failure`, just like in `loki.service`?
